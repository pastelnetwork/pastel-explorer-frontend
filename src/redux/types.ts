import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';
import { RootState } from '@redux/store';

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export type AppThunkDispatch = ThunkDispatch<RootState, unknown, Action<string>>;
