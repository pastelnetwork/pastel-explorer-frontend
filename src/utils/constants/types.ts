import { LabelKeyObject } from 'react-csv/components/CommonPropTypes';
import { PeriodTypes, TGranularity } from '@utils/helpers/statisticsLib';
import { TStatisticsInfo } from '@utils/types/IStatistics';

export type TThemeColor = {
  name: string;
  backgroundColor: string;
  lineColor?: string;
  splitLineColor: string;
  color: string;
  stack?: string;
  smooth?: boolean;
};

export type TThemeInitOption = {
  theme?: TThemeColor | null;
  data?: number[][];
  dataX?: string[];
  dataY?: number[];
  dataY1?: number[];
  dataY2?: number[];
  chartName: string;
  minY: number;
  maxY: number;
};

export type TLineChartProps = {
  chartName: string;
  period?: PeriodTypes;
  dataX?: string[];
  dataY?: number[];
  dataY1?: number[];
  dataY2?: number[];
  title?: string;
  granularity?: TGranularity;
  info: TStatisticsInfo;
  granularities?: TGranularity[];
  offset: number;
  periods?: PeriodTypes[];
  handleGranularityFilterChange?: (_granularity: TGranularity) => void;
  handleBgColorChange: (_color: string) => void;
  handlePeriodFilterChange?: (_period: PeriodTypes) => void;
};

export type TCsvHeaderType = {
  [key: string]: LabelKeyObject[];
};

export type TScatterChartProps = {
  chartName: string;
  data: number[][];
  dataX: string[];
  title?: string;
  period?: PeriodTypes;
  info: TStatisticsInfo;
  offset: number;
  periods: PeriodTypes[];
  handleBgColorChange: (_color: string) => void;
  handlePeriodFilterChange?: (_period: PeriodTypes) => void;
  handleGranularityFilterChange?: (_granularity: TGranularity) => void;
};
