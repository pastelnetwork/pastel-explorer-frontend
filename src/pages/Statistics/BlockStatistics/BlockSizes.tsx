import { useState, ChangeEvent, useEffect } from 'react';
import { Skeleton } from '@material-ui/lab';
import { format, fromUnixTime } from 'date-fns';
import parse from 'html-react-parser';

import { PeriodTypes, generatePeriodToDropdownOptions } from '@utils/helpers/statisticsLib';
import { periods, cacheList } from '@utils/constants/statistics';
import { LineChart } from '@components/Summary/LineChart';
import { Dropdown } from '@components/Dropdown/Dropdown';
import { useDeferredData } from '@utils/helpers/useFetch/useFetch';
import { IBlock } from '@utils/types/IBlocks';
import * as URLS from '@utils/constants/urls';
import { translate } from '@utils/helpers/i18n';
import { readCacheValue, setCacheValue } from '@utils/helpers/localStorage';

import * as SummaryStyles from '@components/Summary/Summary.styles';
import * as Styles from '@pages/CascadeAndSenseStatistics/CascadeAndSenseStatistics.styles';
import * as StatisticsStyles from '../Statistics.styles';

import { ITransformBlocksData } from './BlockStatistics.helpers';

interface ChartProps {
  labels: Array<string>;
  data: Array<number>;
}

interface IBlockSizes {
  blockElements: ITransformBlocksData[];
}

const BlockSizes: React.FC<IBlockSizes> = ({ blockElements }) => {
  const [period, setPeriod] = useState<PeriodTypes>(periods[2][0]);
  const [chartData, setChartData] = useState<ChartProps | null>(null);
  const [isLoading, setLoading] = useState(false);
  const generateChartData = (blocks: Array<IBlock>) => {
    const groupedBlocks = blocks.reduce(
      (acc: ChartProps, { size, timestamp }) => {
        const time = format(fromUnixTime(timestamp), 'MM/dd/yyyy hh:mm aa');

        acc.labels.push(time);
        acc.data.push(size / 1024);

        return acc;
      },
      { labels: [], data: [] },
    );

    return groupedBlocks;
  };

  const blockSizesData = useDeferredData<{ data: Array<IBlock> }, ChartProps>(
    { method: 'get', url: `${URLS.BLOCK_SIZE_URL}?period=${period}` },
    ({ data }) => generateChartData(data.sort((a, b) => a.timestamp - b.timestamp)),
    undefined,
    undefined,
    [period, blockElements],
  );

  useEffect(() => {
    let currentCache = readCacheValue(cacheList.blockSizes) || {};
    if (currentCache[period]) {
      setChartData(currentCache[period].parseData as ChartProps);
      setLoading(false);
    } else {
      setLoading(true);
    }
    if (!blockSizesData.isLoading && blockSizesData.data) {
      const parseData = blockSizesData.data;
      setChartData(parseData);
      currentCache = {
        ...currentCache,
        [period]: {
          parseData,
        },
      };
      setCacheValue(
        cacheList.blockSizes,
        JSON.stringify({
          currentCache,
          lastDate: Date.now(),
        }),
      );
      setLoading(false);
    }
  }, [period, blockSizesData.isLoading]);

  const handleDropdownChange = (
    event: ChangeEvent<{
      value: unknown;
    }>,
  ) => {
    if (event.target.value) {
      setPeriod(event.target.value as PeriodTypes);
    }
  };

  return (
    <SummaryStyles.Card className="cascade-sense-card">
      <SummaryStyles.CardContent>
        <SummaryStyles.ValueWrapper>
          <SummaryStyles.Typography variant="h6">
            {parse(translate('pages.statistics.blockSizes'))}
          </SummaryStyles.Typography>
        </SummaryStyles.ValueWrapper>
        <SummaryStyles.PercentageWrapper>
          <Styles.Percentage>
            <Dropdown
              value={period}
              onChange={handleDropdownChange}
              options={generatePeriodToDropdownOptions(periods[2])}
              classNameWrapper="cascade-sense-statistics"
            />
          </Styles.Percentage>
        </SummaryStyles.PercentageWrapper>
      </SummaryStyles.CardContent>
      <StatisticsStyles.ChartSection>
        {isLoading ? (
          <StatisticsStyles.Loader>
            <StatisticsStyles.LoadingText>
              {parse(translate('common.loadingData'))}
            </StatisticsStyles.LoadingText>
          </StatisticsStyles.Loader>
        ) : null}
        {!chartData ? (
          <>
            <Skeleton animation="wave" variant="rect" height={300} />
            <StatisticsStyles.LoadingText>
              {parse(translate('common.loadingData'))}
            </StatisticsStyles.LoadingText>
          </>
        ) : (
          <LineChart
            chartName="blockSizesStatistics"
            dataX={chartData?.labels}
            dataY={chartData?.data}
            offset={1000}
            disableClick
          />
        )}
      </StatisticsStyles.ChartSection>
    </SummaryStyles.Card>
  );
};

export default BlockSizes;
