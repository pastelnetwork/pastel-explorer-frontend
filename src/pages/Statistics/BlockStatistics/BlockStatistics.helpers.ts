import { getCurrentUnixTimestamp } from '@utils/helpers/date/date';
import { formatNumber } from '@utils/helpers/formatNumbers/formatNumbers';
import { IBlock } from '@utils/types/IBlocks';

import themeVariant from '@theme/variants';

export interface ITransformBlocksData {
  id: string;
  transactionCount: string;
  height: string;
  size: string;
  minutesAgo: string;
}

export const transformBlocksData = (
  transactions: Array<IBlock>,
  currentTime: number,
): Array<ITransformBlocksData> => {
  return transactions.map(({ id, timestamp, transactionCount, height, size }) => {
    const minutesAgo = Math.floor(((currentTime || getCurrentUnixTimestamp) - timestamp) / 60);
    return {
      id,
      transactionCount: `${transactionCount} transaction${transactionCount === 1 ? '' : 's'}`,
      height: formatNumber(height),
      size: `${(size / 1024).toFixed(2)} kB`,
      minutesAgo: `${minutesAgo} minute${minutesAgo === 1 ? '' : 's'} ago`,
    };
  });
};

export const generateBlocksChartData = (labels: Array<string>, data: Array<number>) => ({
  labels,
  datasets: [
    {
      label: 'Block KB',
      data,
      fill: false,
      borderColor: themeVariant.map.peer,
      tension: 0.1,
      pointRadius: 2,
      borderWidth: 1,
    },
  ],
});
