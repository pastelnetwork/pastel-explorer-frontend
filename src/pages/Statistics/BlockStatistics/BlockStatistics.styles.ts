import styled from 'styled-components/macro';
import { Grid } from '@material-ui/core';

export const BlocksContainer = styled(Grid)`
  flex-wrap: nowrap;
  // max-height: 172px;
  padding: 25px 20px;
  overflow-x: scroll;
`;
