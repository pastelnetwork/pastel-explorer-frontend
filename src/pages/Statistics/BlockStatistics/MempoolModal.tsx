import { makeStyles, createStyles } from '@material-ui/core/styles';
import Modal from '@material-ui/core/Modal';
import Backdrop from '@material-ui/core/Backdrop';
import Fade from '@material-ui/core/Fade';
import Tooltip from '@material-ui/core/Tooltip';
import { Table, TableBody, TableHead, TableRow } from '@material-ui/core';
import CircularProgress from '@material-ui/core/CircularProgress';
import TimeAgo from 'react-timeago';

import { getCurrencyName } from '@utils/appInfo';
import RouterLink from '@components/RouterLink/RouterLink';
import * as ROUTES from '@utils/constants/routes';
import { formatAddress } from '@utils/helpers/format';
import { formatNumber } from '@utils/helpers/formatNumbers/formatNumbers';
import { translate } from '@utils/helpers/i18n';
import { useMempool } from '@hooks/useBlockStatistics';
import { getTicketsTypeList } from '@pages/Movement/Movement.helpers';
import * as TransactionStyles from '@pages/Details/TransactionDetails/TransactionDetails.styles';
import * as TableStyles from '@components/Table/Table.styles';

import * as Styles from '../Statistics.styles';

interface IMempoolModal {
  open: boolean;
  onClose: () => void;
}

const MempoolModal: React.FC<IMempoolModal> = ({ open, onClose }) => {
  const { mempools, isLoading } = useMempool();
  const useStyles = makeStyles(() =>
    createStyles({
      modal: {
        display: 'flex',
        justifyContent: 'center',
        overflowY: 'auto',
        height: '100%',
      },
    }),
  );
  const classes = useStyles();

  const getModalContent = () => {
    if (isLoading) {
      return (
        <TransactionStyles.LoadingWrapper>
          <TransactionStyles.Loader>
            <CircularProgress size={40} />
          </TransactionStyles.Loader>
        </TransactionStyles.LoadingWrapper>
      );
    }

    return (
      <TableStyles.BlockWrapper className="mempool-wrapper">
        <TableStyles.BlockTitle>{translate('pages.movement.mempool')}</TableStyles.BlockTitle>
        <Styles.ContentWrapper>
          <TableStyles.PaperWrapper>
            <TableStyles.TableWrapper>
              <Table className="custom-table mempool-table">
                <TableHead className="table__row-header">
                  <TableRow>
                    <TableStyles.TableCell className="table__row-header--cell">
                      {translate('pages.movement.txID')}
                    </TableStyles.TableCell>
                    <TableStyles.TableCell className="table__row-header--cell">
                      {translate('pages.movement.recipients')}
                    </TableStyles.TableCell>
                    <TableStyles.TableCell className="table__row-header--cell">
                      {translate('pages.movement.amount', { currency: getCurrencyName() })}
                    </TableStyles.TableCell>
                    <TableStyles.TableCell className="table__row-header--cell">
                      {translate('pages.movement.size')}
                    </TableStyles.TableCell>
                    <TableStyles.TableCell className="table__row-header--cell">
                      {translate('pages.movement.fee', { currency: getCurrencyName() })}
                    </TableStyles.TableCell>
                    <TableStyles.TableCell className="table__row-header--cell">
                      {translate('pages.movement.ticketQuantity')}
                    </TableStyles.TableCell>
                    <TableStyles.TableCell className="table__row-header--cell">
                      {translate('pages.movement.timestamp')}
                    </TableStyles.TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {mempools?.map(mempool => {
                    const ticketsTypeList = getTicketsTypeList(mempool.tickets || '');
                    return (
                      <TableRow className="table__row" key={mempool.id}>
                        <TableStyles.RowCell
                          data-title={`${translate('pages.movement.txID')}:`}
                          className="cell-content"
                        >
                          <RouterLink
                            route={`${ROUTES.TRANSACTION_DETAILS}/${mempool.id}`}
                            value={formatAddress(mempool.id)}
                            title={mempool.id}
                            className="address-link"
                          />
                        </TableStyles.RowCell>
                        <TableStyles.RowCell
                          data-title={`${translate('pages.movement.recipients')}:`}
                          className="cell-content"
                        >
                          {formatNumber(mempool.recipientCount)}
                        </TableStyles.RowCell>
                        <TableStyles.RowCell
                          data-title={`${translate('pages.movement.amount', {
                            currency: getCurrencyName(),
                          })}:`}
                          className="cell-content"
                        >
                          {mempool.isNonStandard ? (
                            <Tooltip title={translate('pages.movement.shieldedTransactionInfo')}>
                              <span>{translate('common.unknown')}</span>
                            </Tooltip>
                          ) : (
                            formatNumber(mempool.totalAmount, { decimalsLength: 2 })
                          )}
                        </TableStyles.RowCell>
                        <TableStyles.RowCell
                          data-title={`${translate('pages.movement.size')}:`}
                          className="cell-content"
                        >
                          {formatNumber(mempool.size / 1024, { decimalsLength: 2 })}{' '}
                          {translate('pages.movement.kb')}
                        </TableStyles.RowCell>
                        <TableStyles.RowCell
                          data-title={`${translate('pages.movement.fee', {
                            currency: getCurrencyName(),
                          })}:`}
                          className="cell-content"
                        >
                          {formatNumber(mempool.fee, { decimalsLength: 2 })}
                        </TableStyles.RowCell>
                        <TableStyles.RowCell
                          data-title={`${translate('pages.movement.ticketQuantity')}:`}
                          className="cell-content"
                        >
                          {ticketsTypeList.total > 0 ? (
                            <RouterLink
                              route={`${ROUTES.TRANSACTION_DETAILS}/${mempool.id}`}
                              value={ticketsTypeList.total}
                              textSize="large"
                              title={ticketsTypeList.text.join(', <br />')}
                              isUseTooltip
                            />
                          ) : (
                            0
                          )}
                        </TableStyles.RowCell>
                        <TableStyles.RowCell
                          data-title={`${translate('pages.movement.timestamp')}:`}
                          className="cell-content"
                        >
                          <TimeAgo date={mempool.timestamp * 1000} />
                        </TableStyles.RowCell>
                      </TableRow>
                    );
                  })}
                  {!mempools?.length ? (
                    <TableRow className="table__row">
                      <TableStyles.RowCell className="cell-content" colSpan={8}>
                        <Styles.NoContent>{translate('common.noData')}</Styles.NoContent>
                      </TableStyles.RowCell>
                    </TableRow>
                  ) : null}
                </TableBody>
              </Table>
            </TableStyles.TableWrapper>
          </TableStyles.PaperWrapper>
        </Styles.ContentWrapper>
      </TableStyles.BlockWrapper>
    );
  };
  return (
    <Modal
      aria-labelledby="transition-modal-title"
      aria-describedby="transition-modal-description"
      className={classes.modal}
      open={open}
      onClose={onClose}
      closeAfterTransition
      BackdropComponent={Backdrop}
      BackdropProps={{
        timeout: 500,
      }}
    >
      <Fade in={open}>
        <Styles.ModalContent>{getModalContent()}</Styles.ModalContent>
      </Fade>
    </Modal>
  );
};

export default MempoolModal;
