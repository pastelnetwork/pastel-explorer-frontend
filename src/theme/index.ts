import { createMuiTheme, Theme } from '@material-ui/core/styles';
import themeVariant from './variants';
import typography from './typography';
import overrides from './overrides';
import breakpoints from './breakpoints';
import props from './props';
import shadows from './shadows';

const createTheme = (): Theme => {
  return createMuiTheme(
    {
      spacing: 4,
      breakpoints,
      overrides,
      props,
      typography,
      shadows,
      palette: themeVariant.palette,
    },
    {
      header: themeVariant.header,
      footer: themeVariant.footer,
      sidebar: themeVariant.sidebar,
    },
  );
};

export const themeLight = createMuiTheme(
  {
    palette: {
      ...themeVariant.palette,
      background: {
        default: '#FFFFFF',
        paper: '#FAFBFC',
      },
      text: {
        primary: '#2D3748',
        secondary: '#4E5D78',
      },
    },
    spacing: 4,
    breakpoints,
    overrides,
    props,
    typography,
    shadows,
  },
  {
    header: themeVariant.header,
    footer: themeVariant.footer,
    sidebar: {
      ...themeVariant.sidebar,
      background: {
        default: '#FFFFFF',
        active: '#6C5DD3',
      },
      text: {
        default: '#FFFFFF',
        active: '#F8F8FA',
        primary: '#4A5568',
      },
      radius: {
        default: 0,
        active: '15px',
      },
    },
  },
);

export const themeDark = createMuiTheme(
  {
    palette: {
      ...themeVariant.palette,
      background: {
        default: '#232630',
        paper: '#14161D',
      },
      text: {
        primary: '#F8F8FA',
        secondary: '#F8F8FA',
      },
    },
    spacing: 4,
    breakpoints,
    overrides,
    props,
    typography,
    shadows,
  },
  {
    header: themeVariant.header,
    footer: themeVariant.footer,
    sidebar: {
      ...themeVariant.sidebar,
      background: {
        default: '#232630',
        active: '#6C5DD3',
      },
      text: {
        default: '#4A5568',
        active: '#F8F8FA',
        primary: '#4A5568',
      },
      radius: {
        default: 0,
        active: '15px',
      },
    },
  },
);

export default createTheme;

export type TAppTheme = typeof themeDark;
