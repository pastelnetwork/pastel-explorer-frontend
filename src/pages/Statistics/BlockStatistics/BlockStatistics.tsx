import * as React from 'react';
// third party
import { useHistory } from 'react-router-dom';
import { format, fromUnixTime } from 'date-fns';

import { Grid, darken } from '@material-ui/core';
import { Skeleton } from '@material-ui/lab';
import { makeStyles } from '@material-ui/styles';
import { TAppTheme } from '@theme/index';
// application
import Header from '@components/Header/Header';
import { EChartsLineChart } from '@pages/HistoricalStatistics/Chart/EChartsLineChart';
import { BlockUnconfirmed } from '@utils/types/ITransactions';

import * as ROUTES from '@utils/constants/routes';
import * as URLS from '@utils/constants/urls';
import { useFetch } from '@utils/helpers/useFetch/useFetch';
import { IBlock } from '@utils/types/IBlocks';
import { useBackgroundChart } from '@utils/hooks';

import { info } from '@utils/constants/statistics';
import { SocketContext } from '@context/socket';

import BlockVisualization from './BlockVisualization/BlockVisualization';
import { transformBlocksData, ITransformBlocksData } from './BlockStatistics.helpers';

const BLOCK_ELEMENTS_COUNT = 8;

const useStyles = makeStyles((_theme: TAppTheme) => ({
  wrapper: {
    margin: 0,
    [_theme.breakpoints.up('md')]: {
      width: 'calc(100vw - 314px)',
    },
  },
  root: {
    padding: '25px 20px',
    overflowX: 'auto',
    width: '100%',
    margin: 0,
    marginBottom: 16,
    '&::-webkit-scrollbar': {
      background: _theme.palette.background.default,
    },
    '&::-webkit-scrollbar-thumb': {
      background: darken(_theme.palette.background.paper, 0.5),
    },
  },
}));

interface ChartProps {
  labels: Array<string>;
  data: Array<number>;
}

const StatisticsBlocks: React.FC = () => {
  const history = useHistory();
  const classes = useStyles();
  const [blockElements, setBlockElements] = React.useState<Array<ITransformBlocksData>>([]);
  const [chartData, setChartData] = React.useState<ChartProps | null>(null);
  const [currentBgColor, handleBgColorChange] = useBackgroundChart();
  const [blocksUnconfirmed, setBlocksUnconfirmed] = React.useState<BlockUnconfirmed[] | null>(null);
  const fetchUnconfirmedTxs = useFetch<{ data: BlockUnconfirmed[] }>({
    method: 'get',
    url: URLS.GET_UNCONFIRMED_TRANSACTIONS,
  });
  const socket = React.useContext(SocketContext);

  const fetchBlocksData = useFetch<{ data: Array<IBlock>; timestamp: number }>({
    method: 'get',
    url: URLS.BLOCK_URL,
  });

  const generateChartData = (blocks: Array<IBlock>) => {
    const groupedBlocks = blocks.reduce(
      (acc: ChartProps, { size, timestamp }) => {
        const time = format(fromUnixTime(timestamp), 'HH:mm');

        acc.labels.push(time);
        acc.data.push(size / 1024);

        return acc;
      },
      { labels: [], data: [] },
    );

    return groupedBlocks;
  };
  const handleBlocksData = () => {
    Promise.all([
      fetchBlocksData.fetchData({ params: { limit: BLOCK_ELEMENTS_COUNT } }),
      fetchUnconfirmedTxs.fetchData(),
    ]).then(([blocksData, txs]) => {
      if (blocksData) {
        setChartData(generateChartData(blocksData.data));
        setBlockElements(transformBlocksData(blocksData.data, blocksData.timestamp));
      }
      setBlocksUnconfirmed(txs?.data || null);
    });
  };
  React.useEffect(() => {
    socket.on('getUpdateBlock', () => {
      handleBlocksData();
    });
    return () => {
      socket.off('getUpdateBlock');
    };
  }, []);
  React.useEffect(() => {
    handleBlocksData();
  }, []);

  return (
    <>
      <Header title="Blocks Statistics" />
      <Grid classes={{ root: classes.wrapper }} container>
        {blockElements && blockElements.length ? (
          <Grid
            classes={{ container: classes.root }}
            className={classes.root}
            wrap="nowrap"
            container
            justify="flex-start"
            alignItems="center"
            spacing={8}
          >
            {blocksUnconfirmed && blocksUnconfirmed.length
              ? blocksUnconfirmed.map(({ height, size, txsCount }, idx) => (
                  <Grid item key={height}>
                    <BlockVisualization
                      height="--"
                      className="block-unconfirmed"
                      size={`${(size / 1024).toFixed(2)} kB`}
                      transactionCount={`${txsCount} transactions`}
                      minutesAgo={`In ~${(blocksUnconfirmed.length - idx) * 10} minutes`}
                    />
                  </Grid>
                ))
              : null}
            <Grid item>
              <div
                style={{
                  width: 3,
                  height: 145,
                  borderLeft: '3px dashed',
                  marginLeft: -10,
                  marginTop: -30,
                }}
              />
            </Grid>
            {blockElements.slice(1, 8).map(({ id, height, size, transactionCount, minutesAgo }) => (
              <Grid item key={id}>
                <BlockVisualization
                  clickHandler={() => history.push(`${ROUTES.BLOCK_DETAILS}/${id}`)}
                  height={height}
                  size={size}
                  transactionCount={transactionCount}
                  minutesAgo={minutesAgo}
                />
              </Grid>
            ))}
          </Grid>
        ) : (
          <Skeleton animation="wave" variant="rect" height={300} />
        )}
        <Grid item xs={12} lg={12}>
          {chartData && (
            <div style={{ flex: 1, backgroundColor: currentBgColor }}>
              <EChartsLineChart
                chartName="hashrate"
                dataX={chartData?.labels}
                dataY={chartData?.data}
                title="Block sizes (kB)"
                info={info}
                offset={1}
                handleBgColorChange={handleBgColorChange}
              />
            </div>
          )}
        </Grid>
      </Grid>
    </>
  );
};

export default StatisticsBlocks;
