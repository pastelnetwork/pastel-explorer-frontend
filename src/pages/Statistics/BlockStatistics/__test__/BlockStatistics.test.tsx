import { shallow } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import 'jest-styled-components';

import { MyMockType } from '@utils/types/MockType';
import i18next from '../../../../utils/helpers/i18n';
import { themeLight } from '../../../../theme';
import BlockStatistics from '../BlockStatistics';

jest.mock('gsap');
jest.mock('i18next-http-backend');
jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            // noop
          }),
      },
    };
  },
  initReactI18next: {
    type: '3rdParty',
    init: () => {
      // noop
    },
  },
}));
i18next.t = jest.fn().mockImplementation((...arg) => {
  return arg[0];
}) as MyMockType;

describe('pages/Statistics/BlockStatistics', () => {
  const blockElements = [
    {
      id: '245684',
      transactionCount: '3',
      height: '245684',
      size: '2.5kB',
      minutesAgo: '3 minutes',
      ticketsCount: '0',
      status: 'entered',
    },
  ];
  const blocksUnconfirmed = null;
  const wrapper = shallow(
    <ThemeProvider theme={themeLight}>
      <BlockStatistics blockElements={blockElements} blocksUnconfirmed={blocksUnconfirmed} />,
    </ThemeProvider>,
  );

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
