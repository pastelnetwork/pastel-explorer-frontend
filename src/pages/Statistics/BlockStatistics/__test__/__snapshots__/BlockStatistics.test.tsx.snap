// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Statistics/BlockStatistics renders correctly 1`] = `ShallowWrapper {}`;
